### PR TITLE
Add Basic Authentication support

### PR DESCRIPTION
**Pull request by Chamindu36**
## Purpose
Provide authentication for product micro-gateway with basic authentication which uses usernames and passwords for authentication.

## Goals
Providing feature of enabling basic authentication security schema to product micro-gateway along with OAuth2 token based authentication. User will be benefited with using only OAuth2 token based authentication alone, using basic authentication alone and use both schemas at same time. 

## Approach
The ballerina source of the current micro-gateway is generated only for supporting OAuth2.0 token based authentication. We will introduce a new setup command flag as “-b” and “--security” which can be configured at  'setup' command as “--security=”values_of_securitySchemas”” to make the micro-gateway perform  based on basic authentication and as well as work with OAuth 2.0. When this command is provided, the ballerina source will be generated to support following user stories based on what are the values user is provided.Then at “service.mustache”  file, it has to check whether this isOptional and isRequired boolean values at “BasicAuth.java” class which has to be created under “Config.java” . Based on that value the source generation will be changed as follows at “service.mustache” file. Introduced a new file called “BasicAuth.bal” which perform the extraction of credentials from the request, validation and value settings for analytics inside it. Include this filter by default to the filter chain and if “AUTH_HEADER” is “basic” then request will process through this “BasicAuth.bal”. By using this design, this feature can provide support for basic authentication and OAuth2 based authentication at the same time and to only support for basic authentication only.
For further reference :
https://docs.google.com/document/d/11gMFm9yHlnDz2VXRBFSBCHjoLH3-F5w3zXpOaZG0ygc/edit?usp=sharing

## User stories

User Story | Micro-gateway that is only secured with oauth2.
-- | --
Overview | As an user, I need to secure my API only with oauth2.
Persona | API creator, API Publisher
Prerequisites | None
Narrative | User setup micro-gateway.Then user build micro-gateway and no additional steps are needed.Or user can setup micro gateway with --security ”oauth2” command
Acceptance Criteria | Micro-gateway that is secured with oauth2 token should be allowed to access, only if token validation passesMicro-gateway secured with this Oauth2 cannot be validated usernames and passwords.

User Story | Micro-gateway that is only secured with basic authentication.
-- | --
Overview | As an API creator/API Publisher,I need to secure my API only with usernames and passwords.
Persona | API creator, API Publisher
Prerequisites | Passwords converted into their SHA1 hash values.
Narrative | User setup micro-gateway with following flag and value.--security ”basic” Then user build micro-gateway and no additional steps are needed.User has to input authorized usernames and passwords to micro-gw.conf file.
Acceptance Criteria | Micro-gateway that is secured with basic auth should be allowed to access with authorized usernames and passwords.Micro-gateway with basic authentication should not be allowed to access with a valid oauth2 token.

User Story | Micro-gateway which is secured with basic authentication and with oauth2 token on same time.
-- | --
Overview | User need to secure micro-gateway with both oauth2 token or with usernames and passwords.
Persona | API creator, API Publisher
Prerequisites | Passwords converted into their SHA1 hash values.
Narrative | User setup micro-gateway with following flag and value.           --security ”oauth2,basic” Then user build micro-gateway and no additional steps are needed.User has to input authorized usernames and passwords to micro-gw.conf file.
Acceptance Criteria | Micro-gateway that is secured with both should be allowed to access with authorized usernames and passwords.Micro-gateway that is secured with both schemas should be allowed to access, only if token validation passes

## Documentation
Doc has Impact

## Automation tests
 - Unit tests 
testApiInvokeFailWithInvalidFormat
testApiInvokeFailWithInvalidPassword
testApiInvokeWithoutPassword
testApiInvokePassWithBasicAuth
testApiInvokeFailWithJWT
testApiInvokeWithJWT
testApiInvokeWithBasicAuth
testApiInvokeFailWithBasicAuth

 - Integration tests
1. APIInvokeWithBasicAuthTestCase
2. APIInvokeWithOAuth2andBasicAuthTestCase
3. APIInvokeWithOAuthTestCase

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/config/BasicAuth.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    public boolean getOptional() {
+        boolean optional;
+        if (isOptional == true) {
+            optional = true;
+        } else {
+            optional = false;
+        }
+        return optional;
+
```
Comment (npamudika) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/config/BasicAuth.java]: Shall we remove this additional line?

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenApiCodegenUtils.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            throw new CLIInternalException("Error while reading api definition.");
+        }
+        return responseStr;
+    }
+
+    public static void setAdditionalConfigs(ExtendedAPI api) throws IOException {
+        String endpointConfig = api.getEndpointConfig();
+        api.setEndpointConfigRepresentation(getEndpointConfig(endpointConfig));
+    }
+
```
Comment (npamudika) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenApiCodegenUtils.java]: Shall we remove this additional line?

File: components/micro-gateway-core/src/main/ballerina/gateway/constants/api_security_constants.bal [hunk 3]
```diff
@@ -14,96 +14,121 @@
...
+"Make sure you have given the correct username and password";
+
+
+@final public string API_AUTH_BASICAUTH_INVALID_FORMAT_STRING = "900912";
+@final public int API_AUTH_BASICAUTH_INVALID_FORMAT= 900912;
+@final public string API_AUTH_BASICAUTH_INVALID_FORMAT_STRING_MESSAGE = "Invalid Format";
+@final public string API_AUTH_BASICAUTH_INVALID_FORMAT_STRING_DESCRIPTION =
+"Make sure you have given the credentials in correct format with \":\" character";
+
+
```
Comment (npamudika) [inline components/micro-gateway-core/src/main/ballerina/gateway/constants/api_security_constants.bal]: Shall we remove this additional line?

File: components/micro-gateway-core/src/main/ballerina/gateway/listeners/api_gateway_listener.bal [hunk 4]
```diff
@@ -137,47 +138,53 @@ public function getAuthProviders() returns http:AuthProvider[] {
...
+}
+
+public function getBasicAuthProvider() returns http:AuthProvider[] {
+    http:AuthProvider basicAuthProvider = {
+        id: AUTHN_SCHEME_BASIC,
+        scheme: AUTHN_SCHEME_BASIC,
+        authStoreProvider: AUTH_PROVIDER_CONFIG
+    };
+    return [basicAuthProvider];
+}
```
Comment (npamudika) [inline components/micro-gateway-core/src/main/ballerina/gateway/listeners/api_gateway_listener.bal]: Lets add an extra line after the method

File: distribution/resources/conf/micro-gw.conf [hunk 5]
```diff
@@ -57,4 +57,10 @@ ciphers=["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA","TLS_ECDHE_ECDSA_WITH_AES_128_CBC_
...
   "SSL_DHE_DSS_WITH_3DES_EDE_CBC_SHA"," TLS_EMPTY_RENEGOTIATION_INFO_SCSV"]
-sslVerifyClient="not_require"
\ No newline at end of file
+sslVerifyClient="not_require"
+
+["b7a.users"]
+["b7a.users.generalUser1"]
+password="5BAA61E4C9B93F3F0682250B6CF8331B7EE68FD8"
+
+
```
Comment (npamudika) [inline distribution/resources/conf/micro-gw.conf]: Shall we remove this additional line?

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        headers.put(HttpHeaderNames.AUTHORIZATION.toString(), "Bearer " + token);
+        org.wso2.micro.gateway.tests.util.HttpResponse response = HttpClientRequest
+                .doGet(getServiceURLHttp("/pizzashack/1.0.0/menu"), headers);
+        Assert.assertNotNull(response);
+        Assert.assertEquals(response.getResponseCode(), responseCode, "Response code mismatched");
+    }
+
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokePassWithBasicAuth() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Shall we remove this additional line?

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        //test endpoint
+        invokeBasic(basicAuthToken, MockHttpServer.SAND_ENDPOINT_RESPONSE, 200);
+
+        try {
+            Thread.sleep(2000);
+        } catch(InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Shall we remove this additional line?

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+    }
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Insert an extra line here

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokeWithoutPassword() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Shall we remove this additional line?

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        //test endpoint
+        invokeBasic(basicAuthToken, 401);
+
+        try {
+            Thread.sleep(2000);
+        } catch(InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Shall we remove this additional line?

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokeFailWithInvalidPassword() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Shall we remove this additional line?

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        //test endpoint
+        invokeBasic(basicAuthToken, 401);
+
+        try {
+            Thread.sleep(2000);
+        } catch(InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Here too

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    @Test(description = "Test API invocation with Basic Auth")
+    public void testApiInvokeFailWithInvalidFormat() throws Exception {
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Here as well

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+        //test endpoint
+        invokeBasic(basicAuthToken, 401);
+
+        try {
+            Thread.sleep(2000);
+        } catch(InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: This too

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+    private void invokeBasic(String token, String responseData, int responseCode) throws Exception {
+        Map<String, String> headers = new HashMap<>();
+        //test endpoint with token
+        headers.put(HttpHeaderNames.AUTHORIZATION.toString(), "Basic " + token);
+        org.wso2.micro.gateway.tests.util.HttpResponse response = HttpClientRequest
+                .doGet(getServiceURLHttp("/pizzashack/1.0.0/menu"), headers);
+        Assert.assertNotNull(response);
+        Assert.assertEquals(response.getData(), responseData);
+        Assert.assertEquals(response.getResponseCode(), responseCode, "Response code mismatched");
+    }
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Insert an extra line here

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+    private void invokeBasic(String token, int responseCode) throws Exception {
+        Map<String, String> headers = new HashMap<>();
+        //test endpoint with token
+        headers.put(HttpHeaderNames.AUTHORIZATION.toString(), "Basic " + token);
+        org.wso2.micro.gateway.tests.util.HttpResponse response = HttpClientRequest
+                .doGet(getServiceURLHttp("/pizzashack/1.0.0/menu"), headers);
+        Assert.assertNotNull(response);
+        Assert.assertEquals(response.getResponseCode(), responseCode, "Response code mismatched");
+    }
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithBasicAuthTestCase.java]: Shall we remove this additional line?

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuthTestCase.java [hunk 17]
```diff
@@ -133,6 +135,34 @@ private void invoke(String token, int responseCode) throws Exception {
...
+
+        //test endpoint
+        invokeBasic(basicAuthToken, 401);
+
+        try {
+            Thread.sleep(2000);
+        } catch (InterruptedException ex) {
+            Assert.fail("thread sleep interrupted!");
+        }
+
```
Comment (npamudika) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuthTestCase.java]: Shall we remove this additional line?

File: tests/src/test/resources/testng.xml [hunk 18]
```diff
@@ -32,8 +32,13 @@
...
         </packages>
         <classes>
-            <class name="org.wso2.micro.gateway.tests.toolkit.APIInvokeTestCase"/>
+            <class name="org.wso2.micro.gateway.tests.toolkit.APIInvokeWithOAuth2andBasicAuthTestCase"/>
+            <class name="org.wso2.micro.gateway.tests.toolkit.APIInvokeWithBasicAuthTestCase"/>
+            <class name="org.wso2.micro.gateway.tests.toolkit.APIInvokeWithOAuthTestCase"/>
             <class name="org.wso2.micro.gateway.tests.toolkit.ThrottlingTestCase"/>
+            <class name="org.wso2.micro.gateway.tests.toolkit.OASAPIInvokeTestCase"/>
+
+
```
Comment (npamudika) [inline tests/src/test/resources/testng.xml]: Shall we remove these 2 additional lines?

File: components/micro-gateway-core/src/main/ballerina/gateway/basicAuth.bal [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        string authHead = request.getHeader(AUTHORIZATION_HEADER);
+        string[] headers = authHead.trim().split("\\s* \\s*");
+        string encodedCredentials = headers[1];
+        var decodedCredentials = encodedCredentials.base64Decode();
+        //Extract username and password from the request
+        string userName;
+        string passWord;
+        match decodedCredentials {
+            string decoded => {
+                if(!decoded.contains(":")){
```
Comment (mushthaq33) [inline components/micro-gateway-core/src/main/ballerina/gateway/basicAuth.bal]: Formatting issue. Lets format the code.

File: components/micro-gateway-core/src/main/ballerina/gateway/basicAuth.bal [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+                    setErrorMessageToFilterContext(context, API_AUTH_BASICAUTH_INVALID_FORMAT);
+                    sendErrorResponse(listener, request, untaint context);
+                    return false;
+                }
+                string[] decodedCred = decoded.trim().split(":");
+                userName = decodedCred[0];
+                if(lengthof decodedCred < 2){
```
Comment (mushthaq33) [inline components/micro-gateway-core/src/main/ballerina/gateway/basicAuth.bal]: Formatting issue. Lets format the code. Lets check for formatting issues in other places as well.

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuth2andBasicAuthTestCase.java [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+    @Test(description = "Test API invocation with a JWT token")
+    public void testApiInvokeWithJWT() throws Exception {
+        //test prod endpoint with jwt token
+        invoke(jwtTokenProd, MockHttpServer.PROD_ENDPOINT_RESPONSE, 200);
+
+        //test sand endpoint
+        invoke(jwtTokenSand, MockHttpServer.SAND_ENDPOINT_RESPONSE, 200);
+
+        try {
+            Thread.sleep(2000);
```
Comment (mushthaq33) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuth2andBasicAuthTestCase.java]: Why do we need a Thread.sleep here?

Comment (Chamindu36) [reply tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuth2andBasicAuthTestCase.java]: It's to get JWT token  expired and assert that in next invoke.

**Pull request merged**